src/databricks/labs/ucx/azure/access.py

```diff
@@ -267,27 +307,62 @@ def create_uber_principal(self, prompts: Prompts):
+            self._update_cluster_policy_with_spn(
+                policy_id,
+                uber_principal.client.client_id,
+                secret_identifier,
+                storage_accounts,
             )
-            self._installation.save(config)
-            self._update_cluster_policy_with_spn(policy_id, storage_account_info, uber_principal, inventory_database)
-            self._update_sql_dac_with_spn(storage_account_info, uber_principal, inventory_database)
-        except PermissionError:
-            self._azurerm.delete_service_principal(uber_principal.client.object_id)
-        logger.info(f"Update UCX cluster policy {policy_id} with spn connection details for storage accounts")
+            self._update_sql_dac_with_spn(uber_principal.client.client_id, secret_identifier, storage_accounts)
+        except (PermissionError, NotFound):
+            logger.error("Failed to create service principal", exc_info=True)
+            self._delete_uber_principal()  # Clean up dangling resources
+            return
+        config.uber_spn_id = uber_principal.client.client_id
+        self._installation.save(config)
+        logger.info(f"Created service principal ({config.uber_spn_id}) with access to used storage accounts.")
+        logger.info(f"Updated UCX cluster policy {policy_id} with spn connection details for storage accounts")
+
+    def _delete_uber_principal(self):
+        config = self._installation.load(WorkspaceConfig)
+        if config.uber_spn_id is not None:
+            used_storage_accounts = self._get_storage_accounts()
+            storage_accounts = []
+            for storage in self._azurerm.storage_accounts():
+                if storage.name in used_storage_accounts:
+                    storage_accounts.append(storage)
+            for storage_account in storage_accounts:
+                try:
+                    self._azurerm.delete_storage_permission(str(storage_account.id), principal_id=config.uber_spn_id)
+                except NotFound:
+                    pass  # Already deleted
+                except PermissionDenied:
+                    logger.error(
+                        f"Missing permissions to delete storage permission for {storage_account.id}", exc_info=True
+                    )
+            try:
+                self._azurerm.delete_service_principal(config.uber_spn_id)
+            except NotFound:
+                pass  # Already deleted
+            except PermissionDenied:
+                logger.error(f"Missing permissions to delete service principal: {config.uber_spn_id}", exc_info=True)
+            secret_identifier = f"secrets/{config.inventory_database}/{self._UBER_PRINCIPAL_SECRET_KEY}"
+            self._revert_sql_dac_with_spn(config.uber_spn_id, secret_identifier, storage_accounts)
+        if config.policy_id is not None:
+            self._revert_cluster_policy(config.policy_id)
+        self._safe_delete_scope(config.inventory_database)
```

Comment:
add a bit more failsafes around deletions - so that DAC deletion timeout (or any other IO error) would still get us to deletion of secret scope or policy. I assume we should also want to expose some sort of `cleanup` command.

Reply:
Created an issue for exposing the a `cleanup` command: https://github.com/databrickslabs/ucx/issues/2361

Could explain your failsafe a bit more? Now, it is indeed a "all or nothing": if anything fails during the creation of the uber SP, we delete everything that has been created. You want different behavior? Like the cluster policy and warehouse config should not trigger the revert?